### ISSUE TITLE
Authentication fail after #271

### ISSUE THREAD

**Issue by saeedseyfi** [strophe/strophejs#283]
Hi,
Strophe can't establish a connection since #271 is merged. It was working with the same settings before.

I'm using WebSocket and here is the log:

send:
```
<open xmlns='urn:ietf:params:xml:ns:xmpp-framing' to='localhost' version='1.0'/>
```
revieve:
```
<open xmlns='urn:ietf:params:xml:ns:xmpp-framing' id='5952572546526895809' version='1.0' xml:lang='en' from='localhost'/>
```
```
<stream:features xmlns:stream='http://etherx.jabber.org/streams'><mechanisms xmlns='urn:ietf:params:xml:ns:xmpp-sasl'><mechanism>PLAIN</mechanism><mechanism>X-OAUTH2</mechanism><mechanism>SCRAM-SHA-1</mechanism></mechanisms></stream:features>
```
send:
```
<auth xmlns='urn:ietf:params:xml:ns:xmpp-sasl' mechanism='X-OAUTH2'>ADkyMGNmYjhkNjQ5M2IxNzI1OTM5MWY5ODcwMjY0ZTc0QGxvY2FsaG9zdAAzNzg3ZTA3ZjNmY2Q2OTNhZTQwMGY3Mjg1ODdiMzdjYg==</auth>
```
revieve:
```
<failure xmlns='urn:ietf:params:xml:ns:xmpp-sasl'><not-authorized/><text xml:lang='en'>Invalid token</text></failure>
```
```
<close xmlns='urn:ietf:params:xml:ns:xmpp-framing'/>
```

Please advise.

**Comment by saeedseyfi** [comment]
@Egorikhin @jcbrand Please advise.

**Comment by jcbrand** [comment]
Looks like your server is offering X-OAUTH2 and then strophe.js chooses it as auth mechanism instead of SHA1.

Strophe chooses it above SHA1 because (like OAUTHBEARER, which is a related auth mechanism) it's given higher priority than SHA1 in the code.

I think this is probably wrong, and that like the EXTERNAL auth mechanism, X-OAUTH2 and OAUTHBEARER should be given lower priority than SHA1. They should likely be the 3 lowest prioritized auth mechanisms.

@guusdk added the OAUTHBEARER auth mechanism, I'd like to hear from him whether there was a good reason why he gave it a higher priority than SHA1.

For now, as a workaround you can disable X-OAUTH2 as auth mechanism on your server.

**Comment by guusdk** [comment]
I didn't give the priorities to much thought, to be honest. I assumed that, if enabled on the server, it'd be preferable at least over PLAIN and DIGEST-MD5, which is why I added it to the top of the list (at the time of the commit that introduced it).

**Comment by jcbrand** [comment]
@saeedseyfi Why did you close the ticket, I don't see a pull request for this?

**Comment by saeedseyfi** [comment]
Sorry for the delay.
I changed the priorities like this:
```
        registerSASLMechanisms: function (mechanisms) {
            this.mechanisms = {};
            mechanisms = mechanisms || [
                Strophe.SASLOAuthBearer,
                Strophe.SASLXOAuth2,
                Strophe.SASLExternal,
                Strophe.SASLAnonymous,
                Strophe.SASLMD5,
                Strophe.SASLPlain,
                Strophe.SASLSHA1
            ];
            mechanisms.forEach(this.registerSASLMechanism.bind(this));
        },
```

But it didn't work, I still get the not-authorized response. Please advise.

**Comment by saeedseyfi** [comment]
Sorry for the delay.
Do the following mechanisms have the ideal priorities?
```
70      SCRAM-SHA1
60      PLAIN
50      DIGEST-MD5
40      ANONYMOUS
30      EXTERNAL
20      OAUTH-2
10      OAUTH-BEARER
```

**Comment by jcbrand** [comment]
Currently DIGEST-MD5 has a higher priority than PLAIN, and it should stay like that, since using it is preferable to using PLAIN (if both are available).

I propose the following:

```
70      SCRAM-SHA1
60      DIGEST-MD5
50      PLAIN
40      OAUTH-BEARER
30      OAUTH-2
20      ANONYMOUS
10      EXTERNAL
```

**Comment by jcbrand** [comment]
Concerning what @guusdk said:

> I assumed that, if enabled on the server, it'd be preferable at least over PLAIN and DIGEST-MD5, which is why I added it to the top of the list (at the time of the commit that introduced it).

This assumes that the client supports OAuth login. Strophe.js supports the mechanism, but that doesn't mean the client properly supports OAuth login, so logging in fails, as @saeedseyfi found out.

If the client developer wants to make OAuth login a higher priority than PLAIN etc. then they can still change the default priorities encoded in Strophe.js.

**Comment by saeedseyfi** [comment]
#290 here is the PR, please review.

**Comment by saeedseyfi** [comment]
@jcbrand pr is ready. #290
Please review.

**Comment by jcbrand** [comment]
Merged, thanks @saeedseyfi
